src/main/kotlin/org/rust/ide/completion/DeriveCompletionProvider.kt

```diff
@@ -0,0 +1,63 @@
+package org.rust.ide.completion
+
+import com.intellij.codeInsight.completion.CompletionParameters
+import com.intellij.codeInsight.completion.CompletionProvider
+import com.intellij.codeInsight.completion.CompletionResultSet
+import com.intellij.codeInsight.lookup.LookupElementBuilder
+import com.intellij.patterns.ElementPattern
+import com.intellij.patterns.PlatformPatterns.psiElement
+import com.intellij.psi.PsiElement
+import com.intellij.util.ProcessingContext
+import org.rust.lang.RustLanguage
+import org.rust.lang.core.parser.RustPsiTreeUtil
+import org.rust.lang.core.psi.RustCompositeElementTypes
+import org.rust.lang.core.psi.RustMetaItemElement
+import org.rust.lang.core.psi.RustTokenElementTypes
+import org.rust.lang.core.resolve.util.RustResolveUtil
+
+class DeriveCompletionProvider : CompletionProvider<CompletionParameters>() {
```

Comment:
Yes, you can't declare a companion object of an object, but you don't need one :) This should to the trick: 

```Kotlin
object DeriveCompletionProvider : CompletionProvider<CompletionParameters>() {
    private val DERIVABLE_TRAITS = ...
    override fun addCompletions(
        parameters: CompletionParameters,
        context: ProcessingContext?,
        result: CompletionResultSet
    ) {
        ...
    }

    val elementPattern: ElementPattern<PsiElement> get() {
        ...
    }
}
```